client/src/main/java/io/pravega/client/stream/ReaderGroupConfig.java

```diff
@@ -220,8 +220,15 @@ private void verifyStartAndEndStreamCuts(final StreamCut startStreamCut, final S
            final Map<Segment, Long> endPositions = endStreamCut.asImpl().getPositions();
            //check offsets for overlapping segments.
            startPositions.keySet().stream().filter(endPositions::containsKey)
-                        .forEach(s -> Preconditions.checkArgument(startPositions.get(s) <= endPositions.get(s),
-                                "Segment offset in startStreamCut should be <= segment offset in endStreamCut."));
+                        .forEach(s -> {
+                            if (startPositions.get(s) == -1) {
+                                Preconditions.checkArgument(endPositions.get(s) == -1,
```

Comment:
We use `-1` to say that we have reached the end of a segment, and we release completed segments at checkpoint. This case seems to be saying that it is forbidden to use two stream cuts for begin and end such that no checkpoint happened between the two. Is that the semantics we are trying to enforce, and if so, isn't that unnecessarily limiting?

Reply:
This check is verified only on the segments that are present in both start  and end `StreamCut`. 
If both start and end `StreamCut`s have the same segment then the following invariants are valid
- if segment's position in start `StreamCut`  is `-1L`  then segment's position in the end `StreamCut` should also be `-1L`
- segment's Positions in start `StreamCut` ( where position is not `-1L`) should be less or equal to the segment's Position in end `StreamCut`, provided the segment position in end `StreamCut` is not `-1L`

Reply:
@fpj Perhaps an easier way to reason about why the above is correct is: "If you are starting at the end of a segment, and the end cut has the same segment, it cannot possibly be any offset other than the end of the segment"